### ISSUE TITLE
Missing required key 'Key' in params

### ISSUE THREAD

**Issue by sammerry** [MathieuLoutre/grunt-aws-s3#64]
Im getting a missing key 'key' error when attempting to upload to s3. In the AWS SDK release notes it mentions this as a validation error. https://aws.amazon.com/releasenotes/JavaScript/6967335344676381

The Error
```
MissingRequiredParameter: Missing required key 'Key' in params
```

My Config
```
          stage: {
            options: {
              bucket: 'my_s3_bucket',
              differential: true,
              params: {
                key: '',
                CacheControl: '3000'
              }
            },
            files: [
              {src: ['dist/**'], cwd: '', action: 'upload'},
            ]
          },

```

**Comment by dlsso** [comment]
Looks like you need to put your s3 key in the empty quotes in `key: '',`.

**Comment by MathieuLoutre** [comment]
Oh, interesting. Maybe I can default it to `''` so there's not need for an empty key?

**Comment by sammerry** [comment]
@dlsso when i try that, It looks like its failing some sort of validation.

I get the following error:
```
Warning: "params" can only be CacheControl, ContentDisposition, ContentEncoding, ContentLanguage, ContentLength, ContentMD5, Expires, GrantFullControl, GrantRead, GrantReadACP, GrantWriteACP, Metadata, ServerSideEncryption, StorageClass, WebsiteRedirectLocation, ContentType Use --force to continue.
```

**Comment by dlsso** [comment]
If I'm understanding you correctly your first error was telling you "key" was required and the second one tells you "key" is not allowed. Something seems wrong there. I don't know much about this stuff so I can't tell you much, but in my setup the key is in options, like so:
```
			options:{
				accessKeyId:your_key_here,
				secretAccessKey:your_other_key_here,
			},
```

**Comment by lipis-z** [comment]
@dlsso  I'm getting the same error and I have the above values there :/

**Comment by MathieuLoutre** [comment]
@sammerry, @lipis-z

I just reread the config, the keys shouldn't go in the `params` object, but in the parent object (`options`) like so:
```js
aws_s3: {
  options: {
    accessKeyId: '<%= aws.AWSAccessKeyId %>', // Use the variables
    secretAccessKey: '<%= aws.AWSSecretKey %>', // You can also use env variables
    differential: true,
    params: {
        CacheControl: '3000'
   }
  },
  staging: {
    options: {
      bucket: 'my-wonderful-staging-bucket',
    },
    files: [
      {expand: true, cwd: 'dist/staging/scripts/', src: ['**'], dest: 'app/scripts/'},
    ]
  }
}
```

Does that solve your problem? Or have I misunderstood something?

**Comment by lipis-z** [comment]
I have something like that and getting the error :(
```javascript
aws_s3: {
  options: {
    accessKeyId: 'XXX',
    secretAccessKey: 'YYYY',
    region: 'eu-west-1'
  },
  staging: {
    options: {
      bucket: 'xxxxx-staging'
    },
    files: [
      {src: '**', cwd: 's3/', stream: true}
    ]
  }
};
```

Maybe I (or you) have to upgrade the aws tools?

(I'm quite a noob when it comes to AWS so anything that could help I would really appreciate it)

**Comment by MathieuLoutre** [comment]
This looks about right. Which error are you getting?

**Comment by lipis-z** [comment]
```
Running "aws_s3:edge" (aws_s3) task
Uploading to https://s3-eu-west-1.amazonaws.com/xxxxxxx-edge/
Fatal error: Failed to upload ebs.zip with bucket xxxxxxx-edge
MissingRequiredParameter: Missing required key 'Key' in params
```

very strange :/

**Comment by MathieuLoutre** [comment]
Which version of `grunt-aws-s3` are you using?Can you try `0.11.x` or `0.10.x`?




I suspect this might not be linked to your config but to one of the latest changes...

**Comment by crisu83** [comment]
I'm having the same issue, I also tried both 0.11.x and 0.10.x and I still get the same error.

**Comment by crisu83** [comment]
I think I found what was causing this issue. At least I got rid of it by adding ```dest: '/'``` to my files object.

**Comment by sammerry** [comment]
@crisu83 the dest fix worked for me, thanks!!

**Comment by MathieuLoutre** [comment]
Ah! This is awesome @crisu83. Do you think you can add this to the docs with a PR?
@lipis-z try this solution ` {src: '**', cwd: 's3/', stream: true, dest: '/'}`

**Comment by lipis-z** [comment]
Thanks guys.. will try in a bit!!`

**Comment by lipis-z** [comment]
So by adding the `dest: '/'` worked..!

(I have another issue though, that I'll file it separately)

**Comment by MathieuLoutre** [comment]
I've just published `0.12.3` which triggers a warning if no dest is specified and expand isn't used. Also added a note to the docs. I'll close for now, feel free to re-open if needed.

**Comment by crisu83** [comment]
Glad to help.
